runtime/src/stakes.rs

```diff
@@ -225,24 +225,58 @@ impl Stakes<StakeAccount> {
-        F: Fn(&Pubkey) -> Option<AccountSharedData>,
+        F: Fn(&Pubkey) -> Option<AccountSharedData> + Sync,
     {
-        let stake_delegations = stakes.stake_delegations.iter().map(|(pubkey, delegation)| {
-            let Some(stake_account) = get_account(pubkey) else {
-                return Err(Error::StakeAccountNotFound(*pubkey));
-            };
-            let stake_account = StakeAccount::try_from(stake_account)?;
-            // Sanity check that the delegation is consistent with what is
-            // stored in the account.
-            if stake_account.delegation() == *delegation {
-                Ok((*pubkey, stake_account))
-            } else {
-                Err(Error::InvalidDelegation(*pubkey))
-            }
-        });
+        // im::HashMap doesn't support rayon so we manually build a temporary vector. Note this is
+        // what std HashMap::par_iter() does internally too.
+        let stake_delegations_vec = stakes.stake_delegations.into_iter().collect::<Vec<_>>();
+        let stake_delegations = stake_delegations_vec
+            .into_par_iter()
+            // We use fold/reduce to aggregate the results, which does a bit more work than calling
+            // collect_vec_list() and then manually inserting into a map, but it does it in background
+            // threads, so effectively it's faster.
+            .try_fold_with(ImHashMap::new(), |mut map, (pubkey, delegation)| {
+                let Some(stake_account) = get_account(&pubkey) else {
+                    return Err(Error::StakeAccountNotFound(pubkey));
+                };
+
+                // Assert that all valid vote-accounts referenced in stake delegations are already
+                // contained in `stakes.vote_account`.
+                let voter_pubkey = &delegation.voter_pubkey;
+                if stakes.vote_accounts.get(voter_pubkey).is_none() {
+                    if let Some(account) = get_account(voter_pubkey) {
+                        if VoteStateVersions::is_correct_size_and_initialized(account.data())
+                            && VoteAccount::try_from(account.clone()).is_ok()
+                        {
+                            error!("vote account not cached: {voter_pubkey}, {account:?}");
+                            return Err(Error::VoteAccountNotCached(*voter_pubkey));
+                        }
+                    }
+                }
+
+                let stake_account = StakeAccount::try_from(stake_account)?;
+                // Sanity check that the delegation is consistent with what is
+                // stored in the account.
+                if stake_account.delegation() == delegation {
+                    map.insert(pubkey, stake_account);
+                    Ok(map)
+                } else {
+                    Err(Error::InvalidDelegation(pubkey))
+                }
+            })
+            .try_reduce_with(|mut a, b| {
+                a.extend(b);
+                Ok(a)
+            })
```

Comment:
Why not `union` here?
https://docs.rs/im/latest/im/struct.HashMap.html#method.union

Also, the smaller one should extend the bigger one; like:
https://github.com/anza-xyz/agave/blob/e56d314df/runtime/src/stakes.rs#L571-L579

Reply:
I'll switch to union, which does that check internally